crates/storage/provider/src/post_state.rs

```diff
@@ -0,0 +1,828 @@
+                // avoid deallocating where possible, but these values should not be written to the
+                // database.
+                continue
+            }
+
+            for (key, value) in storage.storage {
+                let key = H256(key.to_be_bytes());
+                if let Some(entry) = storages_cursor.seek_by_key_subkey(address, key)? {
+                    if entry.key == key {
+                        storages_cursor.delete_current()?;
+                    }
+                }
+
+                if value != U256::ZERO {
+                    storages_cursor.upsert(address, StorageEntry { key, value })?;
+                }
+            }
+        }
+
+        // Write new account state
+        let mut accounts_cursor = tx.cursor_write::<tables::PlainAccountState>()?;
+        for (address, account) in self.accounts.into_iter() {
+            if let Some(account) = account {
+                accounts_cursor.upsert(address, account)?;
+            } else if accounts_cursor.seek_exact(address)?.is_some() {
+                accounts_cursor.delete_current()?;
+            }
+        }
+
+        // Write bytecode
+        let mut bytecodes_cursor = tx.cursor_write::<tables::Bytecodes>()?;
+        for (hash, bytecode) in self.bytecode.into_iter() {
+            bytecodes_cursor.upsert(hash, bytecode)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use reth_db::{
+        database::Database,
+        mdbx::{test_utils, Env, EnvKind, WriteMap},
+        transaction::DbTx,
+    };
+    use std::sync::Arc;
+
+    #[test]
+    fn extend() {
+        let mut a = PostState::new();
+        a.create_account(Address::zero(), Account::default());
+        a.destroy_account(Address::zero(), Account::default());
+        a.finish_transition();
+
+        assert_eq!(a.transitions_count(), 1);
+        assert_eq!(a.changes().len(), 3);
+
+        let mut b = PostState::new();
+        b.create_account(Address::repeat_byte(0xff), Account::default());
+        b.finish_transition();
+
+        assert_eq!(b.transitions_count(), 1);
+        assert_eq!(b.changes.len(), 1);
+
+        let mut c = a.clone();
+        c.extend(b.clone());
+
+        assert_eq!(c.transitions_count(), 2);
+        assert_eq!(c.changes.len(), a.changes.len() + b.changes.len());
+    }
+
+    #[test]
+    fn write_to_db_account_info() {
+        let db: Arc<Env<WriteMap>> = test_utils::create_test_db(EnvKind::RW);
+        let tx = db.tx_mut().expect("Could not get database tx");
+
+        let mut post_state = PostState::new();
+
+        let address_a = Address::zero();
+        let address_b = Address::repeat_byte(0xff);
+
+        let account_a = Account { balance: U256::from(1), nonce: 1, bytecode_hash: None };
+        let account_b = Account { balance: U256::from(2), nonce: 2, bytecode_hash: None };
+        let account_b_changed = Account { balance: U256::from(3), nonce: 3, bytecode_hash: None };
```

Comment:
nit we'll want to add some builder pattern funcs to the Account struct to make making these nicer, i.e. `Account::default().set_balance(3 /* impl Into<U256>, may not be possible due to ruint perhaps? */).nonce(3);`